src/io/overlapped.cr

```diff
@@ -49,100 +50,187 @@ module IO::Overlapped
     end
   end
 
-  @[Extern]
-  struct OverlappedOperation
-    getter overlapped : LibC::WSAOVERLAPPED
+  def self.wait_queued_completions(timeout)
+    overlapped_entries = uninitialized LibC::OVERLAPPED_ENTRY[1]
 
-    def initialize(@overlapped)
+    if timeout > UInt64::MAX
+      timeout = LibC::INFINITE
+    else
+      timeout = timeout.to_u64
+    end
+    result = LibC.GetQueuedCompletionStatusEx(Crystal::EventLoop.iocp, overlapped_entries, overlapped_entries.size, out removed, timeout, false)
+    if result == 0
+      error = WinError.value
+      if timeout && error.wait_timeout?
+        return true
+      else
+        raise IO::Error.from_os_error("GetQueuedCompletionStatusEx", error)
+      end
+    end
+
+    if removed == 0
+      raise IO::Error.new("GetQueuedCompletionStatusEx returned 0")
     end
-  end
 
-  def create_operation
-    overlapped = LibC::WSAOVERLAPPED.new
-    OverlappedOperation.new(overlapped)
+    removed.times do |i|
+      OverlappedOperation.schedule(overlapped_entries[i].lpOverlapped) { |fiber| yield fiber }
+    end
+
+    false
   end
 
-  def get_overlapped_result(socket, operation)
-    flags = 0_u32
-    result = LibC.WSAGetOverlappedResult(socket, pointerof(operation).as(LibC::OVERLAPPED*), out bytes, false, pointerof(flags))
-    if result.zero?
-      error = WinError.wsa_value
-      yield error
+  class OverlappedOperation
+    enum State
+      INITIALIZED
+      STARTED
```

Comment:
In my first implementation, I enqueued the `OverlappedOperation` inside the `start` method. For this, I needed a separate state `STARTED` for this. This is no longer required since the enqueuing only happens when the operation is cancelled.
I think `STARTED`  could be removed.